### PR TITLE
feat(iam): enhancements to password vcenter password expiration

### PR DESCRIPTION
**Pull request by GaryJBlake**
In order to have a good experience with our community, we recommend that you read the [contributing guidelines](https://github.com/vmware-samples/power-validated-solutions-for-cloud-foundation/blob/main/CONTRIBUTING.md) for making a pull request.

**Summary of Pull Request**

Renamed and refactored functions:
- `Get-VcenterPasswordExpiration` from `Get-VCPasswordPolicy`
- `Set-VcenterPasswordExpiration` from `Set-VCPasswordPolicy`
- `Get-VcenterRootPasswordExpiration` from `Get-VCRootPasswordExpiry`
- `Set-VcenterRootPasswordExpiration` from `Set-VCPasswordExpiry`

- Removed function `Get-VCPasswordExpiry` as a duplicate.

- Added `Request-VcenterPasswordExpiration` cmdlet to retrieve the global password expiration policy for vCenter Server.
- Added `Request-VcenterRootPasswordExpiration` cmdlet to retrieve the root user password expiration policy for vCenter Server.
- Added `Update-VcenterPasswordExpiration` cmdlet to update the global password expiration policy for vCenter Server.
- Added `Update-VcenterRootPasswordExpiration` cmdlet to update the root user password expiration policy for vCenter Server.
- Added `Request-SsoPasswordExpiration` cmdlet to retrieve the password expiration policy for vCenter Single Sign-On domain.
- Added `Request-SsoPasswordComplexity` cmdlet to retrieve the password complexity policy for vCenter Single Sign-On domain.
- Added `Request-SsoAccountLockout` cmdlet to retrieve the account lockout policy for vCenter Single Sign-On domain.
- Added `Update-SsoPasswordExpiration` cmdlet to update the password expiration policy for vCenter Single Sign-On domain.
- Added `Update-SsoPasswordComplexity` cmdlet to update the password complexity policy for vCenter Single Sign-On domain.
- Added `Update-SsoAccountLockout` cmdlet to update the account lockout policy for vCenter Single Sign-On domain.

Moved functions to the deprecated section:
- `Get-VCServerPasswordPolicy`
- `Set-vCenterPasswordExpiration`
- `Get-SsoPasswordPolicies`

Signed-off-by: Gary Blake <gblake@vmware.com>

**Type of Pull Request**

- [ ] This is a bug fix.
- [x] This is an enhancement or feature.
- [ ] This is a code style / formatting update.
- [ ] This is a documentation update.
- [ ] This is a refactoring update.
- [ ] This is something else.

**Related to Existing Issues**

Issue Number: N/A

**Test and Documentation Coverage**

- [x] Tests have been completed (for bug fixes / features).
- [ ] Documentation has been added / updated (for bug fixes / features).

**Breaking Changes?**

- [ ] Yes, there are breaking changes.
- [x] No, there are no breaking changes.

### CODE REVIEW

File: PowerValidatedSolutions.psm1 [hunk 1]
```diff
@@ -16373,6 +16243,256 @@ Function Update-EsxiAccountLockout {
...
+	)
+
+	Try {
+        if (Test-VCFConnection -server $server) {
+            if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+                if (Get-VCFWorkloadDomain | Where-Object { $_.name -eq $domain }) {
+                    if (($vcfVcenterDetails = Get-vCenterServerDetail -server $server -user $user -pass $pass -domain $domain)) {
+                        if (Test-vSphereApiConnection -server $($vcfVcenterDetails.fqdn)) {
+                            if (Test-vSphereApiAuthentication -server $vcfVcenterDetails.fqdn -user $vcfVcenterDetails.ssoAdmin -pass $vcfVcenterDetails.ssoAdminPass) {
+                                # $pwdExpirySettings = Get-VcenterPasswordExpiration
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:16395]: Should this comment be removed?

File: PowerValidatedSolutions.psm1 [hunk 2]
```diff
@@ -17288,40 +17408,38 @@ Function Get-VCConfigurationDNS {
 }
 Export-ModuleMember -Function Get-VCConfigurationDNS
 
-Function Get-VCPasswordPolicy {
+Function Get-VcenterPasswordExpiration {
     <#
     .SYNOPSIS
-        Get the global password policy.
+        Retrive the global password expiration policy
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:17414]:
```suggestion
        Retrieve the global password expiration policy
```

File: PowerValidatedSolutions.psm1 [hunk 3]
```diff
@@ -17288,40 +17408,38 @@ Function Get-VCConfigurationDNS {
...
 
         .DESCRIPTION
-        The Get-VCPasswordPolicy cmdlet gets global password policy for the vCenter Server
+        The Get-VcenterPasswordExpiration cmdlet retrives the global password expiration policy for local users of vCenter Server
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:17417]:
```suggestion
        The Get-VcenterPasswordExpiration cmdlet retrieves the global password expiration policy for local users of vCenter Server
```

File: PowerValidatedSolutions.psm1 [hunk 4]
```diff
@@ -17288,40 +17408,38 @@ Function Get-VCConfigurationDNS {
...
 
         .EXAMPLE
-        Get-VCPasswordPolicy
-        This example gets the global password policy of the vCenter Server
+        Get-VcenterPasswordExpiration
+        This example gets the global password expiration policy of the vCenter Server
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:17421]:
```suggestion
        This example retrieves the global password expiration policy of the vCenter Server
```

File: PowerValidatedSolutions.psm1 [hunk 5]
```diff
@@ -17334,103 +17452,75 @@ Function Set-VCPasswordPolicy {
...
 }
-Export-ModuleMember -Function Set-VCPasswordPolicy
+Export-ModuleMember -Function Set-VcenterPasswordExpiration
 
-Function Get-VCRootPasswordExpiry {
+Function Get-VcenterRootPasswordExpiration {
     <#
     .SYNOPSIS
-		Get the vCenter Servver root user password expiry date.
+		Retrive the root user password expiration policy
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:17465]:
```suggestion
		Retrieve the root user password expiration policy
```

File: PowerValidatedSolutions.psm1 [hunk 6]
```diff
@@ -17334,103 +17452,75 @@ Function Set-VCPasswordPolicy {
...
 
         .DESCRIPTION
-        The Get-VCRootPasswordExpiry cmdlet gets password expiration settings for the vCenter Server root account
+        The Get-VcenterRootPasswordExpiration cmdlet retrives the root user password expiration policy from vCenter Server
 
         .EXAMPLE
-        Get-VCRootPasswordExpiry
-        This example gets the password policy of the vCenter Server root account
+        Get-VcenterRootPasswordExpiration
+        This example gets the root user password expiration policy from vCenter Server
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:17472]:
```suggestion
        This example retrieves the root user password expiration policy from vCenter Server
```

File: PowerValidatedSolutions.psm1 [hunk 7]
```diff
@@ -17334,103 +17452,75 @@ Function Set-VCPasswordPolicy {
...
         .SYNOPSIS
         Set the vcenter password expiry date
 
         .DESCRIPTION
-        The Set-VCPasswordExpiry cmdlet configures password expiration settings for the vCenter Server root account
+        The Set-VcenterRootPasswordExpiration cmdlet configures password expiration settings for the vCenter Server root account
 
         .EXAMPLE
-        Set-VCPasswordExpiry -passwordExpires $true -email "admin@rainpole.io" -maxDaysBetweenPasswordChange 91
+        Set-VcenterRootPasswordExpiration -email "admin@rainpole.io" -maxDays 999 -warnDays 14
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:17495]: Much better structure!

File: PowerValidatedSolutions.psm1 [hunk 8]
```diff
@@ -35640,5 +35730,135 @@ Function Set-EsxiPasswordPolicy {
...
+
+	Try {
+		if (Test-Connection -server $server) {
+			if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+				if (Get-VCFWorkloadDomain | Where-Object {$_.name -eq $domain}) {
+					if (($vcfVcenterDetails = Get-vCenterServerDetail -server $server -user $user -pass $pass -domain $domain)) {
+						if (Test-VsphereConnection -server $($vcfVcenterDetails.fqdn)) {
+							if (Test-VsphereAuthentication -server $vcfVcenterDetails.fqdn -user $vcfVcenterDetails.ssoAdmin -pass $vcfVcenterDetails.ssoAdminPass) {
+								Request-vSphereApiToken -Fqdn $vcfVcenterDetails.fqdn -Username $vcfVcenterDetails.ssoAdmin -Password $vcfVcenterDetails.ssoAdminPass -admin | Out-Null
+								$pwdExpirySettings = Get-VCPasswordExpiry
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:35766]: Should this comment be removed?

File: PowerValidatedSolutions.psm1 [hunk 9]
```diff
@@ -16373,6 +16243,256 @@ Function Update-EsxiAccountLockout {
...
+        .DESCRIPTION
+        The Update-VcenterPasswordExpiration cmdlet configures the global password expiration policy of a vCenter Server.
+        The cmdlet connects to SDDC Manager using the -server, -user, and -password values:
+        - Validates that network connectivity and authentication is possible to SDDC Manager
+        - Validates that network connectivity and authentication is possible to vCenter Server
+		- Configures the global password expiration policy
+
+        .EXAMPLE
+        Update-VcenterPasswordExpiration -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01 -maxDays 999 -minDays 0 -warnDays 14
+        This example configures the password expiration settings for the vCenter Server root account to expire after 80 days with email for warning set to "admin@rainpole.io"
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:16375]: I think that this is meant to be 999 days based on the example:

```suggestion
        This example configures the password expiration settings for the vCenter Server root account to expire after 999 days with email for warning set to "admin@rainpole.io"
```

File: PowerValidatedSolutions.psm1 [hunk 10]
```diff
@@ -16373,6 +16243,256 @@ Function Update-EsxiAccountLockout {
...
+        .DESCRIPTION
+        The Update-VcenterRootPasswordExpiration cmdlet configures the root user password expiration policy of a
+        vCenter Server. The cmdlet connects to SDDC Manager using the -server, -user, and -password values:
+        - Validates that network connectivity and authentication is possible to SDDC Manager
+        - Validates that network connectivity and authentication is possible to vCenter Server
+		- Configures the root user password expiration policy
+
+        .EXAMPLE
+        Update-VcenterRootPasswordExpiration -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01 -email "admin@rainpole.io" -maxDays 999 -warnDays 14
+        This example configures the configures password expiration settings for the vCenter Server root account
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:16434]: Suggestion based on the `Set-VcenterRootPasswordExpiration` function.

```suggestion
        This example configures the configures password expiration settings for the vCenter Server root account to e expire after 999 days with email for warning set to "admin@rainpole.io"
```

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Some very minor suggestions for typos and example description.

A couple clarifications if a commented section should be removed.

File: PowerValidatedSolutions.psm1 (hunk 1 above)
Comment (GaryJBlake) [reply PowerValidatedSolutions.psm1:16395]: Fixed

File: PowerValidatedSolutions.psm1 (hunk 2 above)
Comment (GaryJBlake) [reply PowerValidatedSolutions.psm1:17414]: Fixed

File: PowerValidatedSolutions.psm1 (hunk 3 above)
Comment (GaryJBlake) [reply PowerValidatedSolutions.psm1:17417]: Fixed

File: PowerValidatedSolutions.psm1 (hunk 4 above)
Comment (GaryJBlake) [reply PowerValidatedSolutions.psm1:17421]: Fixed

File: PowerValidatedSolutions.psm1 (hunk 10 above)
Comment (GaryJBlake) [reply PowerValidatedSolutions.psm1:16434]: Fixed

File: PowerValidatedSolutions.psm1 (hunk 5 above)
Comment (GaryJBlake) [reply PowerValidatedSolutions.psm1:17465]: Fixed

File: PowerValidatedSolutions.psm1 [hunk 11]
```diff
@@ -17334,103 +17452,75 @@ Function Set-VCPasswordPolicy {
...
-Function Get-VCRootPasswordExpiry {
+Function Get-VcenterRootPasswordExpiration {
     <#
     .SYNOPSIS
-		Get the vCenter Servver root user password expiry date.
+		Retrive the root user password expiration policy
 
         .DESCRIPTION
-        The Get-VCRootPasswordExpiry cmdlet gets password expiration settings for the vCenter Server root account
+        The Get-VcenterRootPasswordExpiration cmdlet retrives the root user password expiration policy from vCenter Server
```
[the remark this answers is not in the document]
Comment (GaryJBlake) [reply PowerValidatedSolutions.psm1:17468]: Fixed

File: PowerValidatedSolutions.psm1 (hunk 6 above)
Comment (GaryJBlake) [reply PowerValidatedSolutions.psm1:17472]: Fixed

File: PowerValidatedSolutions.psm1 (hunk 8 above)
Comment (GaryJBlake) [reply PowerValidatedSolutions.psm1:35766]: Fixed

File: PowerValidatedSolutions.psm1 [hunk 12]
```diff
@@ -16373,6 +16243,616 @@ Function Update-EsxiAccountLockout {
...
+        if (Test-VCFConnection -server $server) {
+            if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+                if (Get-VCFWorkloadDomain | Where-Object { $_.name -eq $domain }) {
+                    if (($vcfVcenterDetails = Get-vCenterServerDetail -server $server -user $user -pass $pass -domain $domain)) {
+                        if (Test-SsoConnection -server $($vcfVcenterDetails.fqdn)) {
+                            if (Test-SsoAuthentication -server $vcfVcenterDetails.fqdn -user $vcfVcenterDetails.ssoAdmin -pass $vcfVcenterDetails.ssoAdminPass) {
+                                if ($SsoPasswordExpiration = Get-SsoPasswordPolicy) {
+                                    $SsoPasswordExpirationObject = New-Object -TypeName psobject
+                                    $SsoPasswordExpirationObject | Add-Member -notepropertyname "Component" -notepropertyvalue "vCenter Single Sign-On"
+                                    $SsoPasswordExpirationObject | Add-Member -notepropertyname "Hostname" -notepropertyvalue $($vcfVcenterDetails.fqdn)
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:16280]: Perhaps this is a little pedantic but should this be FQDN in all occurrences? 

`FQDN = hostname +  domain name`

File: CHANGELOG.md [hunk 13]
```diff
@@ -8,6 +8,24 @@
 - Added `Update-EsxiPasswordExpiration` cmdlet to update the password expiration policy for all ESXi hosts in a cluster.
 - Added `Update-EsxiPasswordComplexity` cmdlet to update the password complexity policy for all ESXi hosts in a cluster.
 - Added `Update-EsxiAccountLockout` cmdlet to update the account lockout policy for all ESXi hosts in a cluster.
+- Added `Request-SsoPasswordExpiration` cmdlet to retrieve the password expiration policy for vCenter Single Sign-On domain.
```
Comment (tenthirtyam) [inline CHANGELOG.md:11]: Example

```suggestion
- Added `Request-SsoPasswordExpiration` foo cmdlet to retrieve the password expiration policy for vCenter Single Sign-On domain.
```

File: PowerValidatedSolutions.psm1 (hunk 12 above)
Comment (tenthirtyam) [reply PowerValidatedSolutions.psm1:16280]: Resolved in f29d4b9f2b1d373bf586413376e2be4521047357

**Review by tenthirtyam — APPROVED** [review]
⭐ ⭐ ⭐ ⭐ ⭐ 

Obviously, these are great additions. The unification of the structure will ensure this is easy to use and understand the outputs.

**Pull request merged**